### PR TITLE
Middleware addendum

### PR DESCRIPTION
**Pull request by basz**
Hi @bakura10 we've talked briefly in a PR somewhere. Here is the result. My hope is you will merge it into the middleware branch on which i based it. It still needs some work, but I think most of the refactoring is done. Locally i can do a password grant request and I get a free token... yeah... all in an expressive app.

I left the configuration - but i believe wiring isn't supposed to be provided with these libraries (it is not a module).

Originally I worked on the zf2 module too, but I stopped with that for now (repo still there). I like the cleanness of this lib compared to some of the others. Hope it can evolve into supporting a oauth2 features. incl jwt.

well please let me know what you think, bas

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -22,7 +22,8 @@
     "zendframework/zend-diactoros": "~1.1",
     "container-interop/container-interop": "^1.1",
     "zendframework/zend-math": "~2.2",
-    "zendframework/zend-eventmanager": "~2.2"
+    "zendframework/zend-eventmanager": "~2.2",
```
Comment (bakura10) [inline composer.json]: v3 is now live and with MUCH better performance :).

File: composer.json [hunk 2]
```diff
@@ -22,7 +22,8 @@
...
+    "zendframework/zend-stdlib": "~2.5"
```
Comment (bakura10) [inline composer.json]: Can't we try to remove this?

File: composer.json [hunk 3]
```diff
@@ -22,7 +22,8 @@
     "zendframework/zend-diactoros": "~1.1",
```
Comment (bakura10) [inline composer.json]: I'd be actually very happy to set min PHP version to PHP7. This would allows us to remove dependency to Zend-Math and performs various simplifications.

File: composer.json [hunk 4]
```diff
@@ -22,7 +22,8 @@
...
     "container-interop/container-interop": "^1.1",
     "zendframework/zend-math": "~2.2",
-    "zendframework/zend-eventmanager": "~2.2"
+    "zendframework/zend-eventmanager": "~2.2",
+    "zendframework/zend-stdlib": "~2.5"
   },
   "require-dev": {
     "phpunit/phpunit": "~4.6",
```
Comment (bakura10) [inline composer.json]: You'll need to upgrade to ~5.1 (esp for PHP 7 compatibility)

File: config/zend-expressive/zend-expressive.config.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
+<?php
```
Comment (bakura10) [inline config/zend-expressive/zend-expressive.config.php]: Actually this config should be agnostic and not tied to zend expressive. You should simply call it "config.global.php". I've already seen bridge for Slim container that consume service manager container, so theorically this could be re-used for any other frameowkr, not just expressive

File: composer.json (hunk 2 above)
Comment (bakura10) [inline composer.json]: Okay I see you're using it for the Options class. I'd really like to remove this useless dependency, maybe you could see if you could not extract the important part of the AbstractOptions class to move it to the options class.

File: config/zend-expressive/zfr-oauth2-server.global.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+use Doctrine\Common\Persistence\ManagerRegistry;
+use ZfrOAuth2\Server\Entity\TokenOwnerInterface;
+
+return [
```
Comment (bakura10) [inline config/zend-expressive/zfr-oauth2-server.global.php]: Not sure separating the config in two files makes much sense.

File: src/Server/Container/AccessTokenServiceFactory.php [hunk 7]
```diff
@@ -29,10 +35,29 @@ class AccessTokenServiceFactory
 {
     /**
      * @param  ContainerInterface $container
-     * @return ClientService
+     * @return TokenService
      */
     public function __invoke(ContainerInterface $container)
```
Comment (bakura10) [inline src/Server/Container/AccessTokenServiceFactory.php]: Note: if we use PHP7 only, return typehint can be added to all factories.

File: src/Server/Grant/PasswordGrant.php [hunk 8]
```diff
@@ -110,8 +110,7 @@ public function createTokenResponse(
             throw OAuth2Exception::invalidRequest('Username and/or password is missing');
         }
 
-        $callback = $this->callback;
-        $owner    = $callback($username, $password);
+        $owner = call_user_func_array($this->callback, [$username, $password]);
```
Comment (bakura10) [inline src/Server/Grant/PasswordGrant.php]: Why this change? Directly calling the method is much more performant than call_user_func_array :).

File: src/Server/Options/ServerOptions.php [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+ * @licence MIT
+ */
+class ServerOptions extends AbstractOptions
+{
+    /**
+     * Object manager key
+     *
+     * @var string|null
+     */
+    protected $objectManager;
```
Comment (bakura10) [inline src/Server/Options/ServerOptions.php]: Is this needed if we use ManagerRegistry?

**Comment by bakura10** [comment]
Thanks a lot @basz !

Just a note for myself: continuing review tomorrow starting from TokenService.

After reading the first part, the only thing I'd like to change is the dependency to Doctrine. I know that @weierophinney would like to use maybe this library for Apigility 2, but Doctrine only is a blocker. Also, I'm using more and more DynamoDB without Doctrine, as I like the lightweight part of it, so I'd definitely want this library to be agnostic: only keeping the code here, maybe using only interfaces for the various repository, and have a Doctrine package that will make the bridge.

We'll need to see how this could be architectured but I'm pretty sure it would be possible.

@weierophinney, what is the expected min version for Apigility 2? I'd love to bump this library to PHP7, but I'd love you to use it for Apigility 2 as well...

File: src/Server/Options/ServerOptions.php (hunk 9 above)
Comment (basz) [inline src/Server/Options/ServerOptions.php]: I think so, though the name suggests an instance while it is in fact something like 'orm_default'. The manager registry takes null or a named service manager.

File: config/zend-expressive/zfr-oauth2-server.global.php (hunk 6 above)
Comment (basz) [inline config/zend-expressive/zfr-oauth2-server.global.php]: Analogue to the zf2 module.config and the often provided module.local.php.dist file. 

But I think we shouldn't provide any configuration at all other then in docs or meta packages.

File: composer.json (hunk 3 above)
Comment (basz) [inline composer.json]: me too.

I noticed a question about apigilities minimum requirement? think you want an answer first?

File: composer.json (hunk 1 above)
Comment (basz) [inline composer.json]: I think it could be "^2.6 || ^3.0" (2.6 introduced intro-container i think)

File: composer.json (hunk 3 above)
Comment (bakura10) [inline composer.json]: Yeah... But I think we can assume PHP7 for now as I'm unsure about when Apigility 2 will be released. So let's assume PHP7+ for now :).

File: composer.json [hunk 10]
```diff
@@ -22,10 +22,10 @@
     "zendframework/zend-diactoros": "~1.1",
     "container-interop/container-interop": "^1.1",
     "zendframework/zend-math": "~2.2",
-    "zendframework/zend-eventmanager": "~2.2"
+    "zendframework/zend-eventmanager": "~2.2 || ~3.0"
```
Comment (bakura10) [inline composer.json]: Actually after some thoughts, event manager is not needed at all in middleware world. The current usage of EventManager was to trigger an event whenever a token is created.

But in middleware this is not needed: you just need to register a middleware just after the main middleware (either at the end of pipeline...) and alter the response. Therefore you can remove this dependency and all usage of event manager in the code :).

File: src/Server/Service/TokenService.php [hunk 11]
```diff
@@ -42,6 +39,11 @@
  */
 class TokenService
 {
+
+    const AUTHORIZATIONCODE_SERVICE = 'ZfrOAuth2\Server\Service\AuthorizationCodeService';
```
Comment (bakura10) [inline src/Server/Service/TokenService.php]: AUTHORIZATION_CODE_SERVICE

File: src/Server/Service/TokenService.php [hunk 12]
```diff
@@ -42,6 +39,11 @@
...
+    const ACCESSTOKEN_SERVICE       = 'ZfrOAuth2\Server\Service\AccessTokenService';
```
Comment (bakura10) [inline src/Server/Service/TokenService.php]: ACCESS_TOKEN_SERVICE

File: src/Server/Service/TokenService.php [hunk 13]
```diff
@@ -42,6 +39,11 @@
...
+    const REFRESHTOKEN_SERVICE      = 'ZfrOAuth2\Server\Service\RefreshTokenService';
```
Comment (bakura10) [inline src/Server/Service/TokenService.php]: REFRESH_TOKEN_SERVICE

File: src/Server/Service/TokenService.php [hunk 14]
```diff
@@ -196,16 +198,16 @@ protected function validateTokenScopes(array $scopes)
      */
     private function compareStrings($expected, $actual)
     {
-        $expected     = (string) $expected;
-        $actual       = (string) $actual;
+        $expected = (string) $expected;
+        $actual   = (string) $actual;
 
         if (function_exists('hash_equals')) {
```
Comment (bakura10) [inline src/Server/Service/TokenService.php]: This is PHP 5.6+, we can assume that we always have it now (in all cases min dependency will be either 5.6 or 7) so the "compareStrings" can be removed and the code only use hash_equals.

**Comment by bakura10** [comment]
Thanks!

So next work: removing all Doctrine dependencies. Using interfaces for repository. For existing factory, you can inject the interface like that:

```php
$foo = $container->get(AccessTokenRepositoryInterface::class);
```

Then each module (like doctrine one) will simply add an alias to the interface to the concrete implementation :).
